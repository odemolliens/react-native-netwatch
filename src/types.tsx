// export type RequestMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
export type RequestMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';

export enum EnumFilterType {
  All = 'ALL',
  Get = 'GET',
  Post = 'POST',
  Put = 'PUT',
  Delete = 'DELETE',
}

export enum EnumStatus {
  Success = 'SUCCESS',
  Warning = 'WARNING',
  Failed = 'FAILED',
}

export type LogType = 'REDUX' | 'RNR' | 'NR' | 'CONNECT';
export type SourceType = 'ALL' | 'RNR' | 'NR' | 'REDUX';

export enum EnumSourceType {
  All = 'ALL',
  ReactNativeRequest = 'RNR',
  Nativerequest = 'NR',
  Redux = 'REDUX',
  Connection = 'CONNECTION',
}

export interface ISourceType {
  format: EnumSourceType | SourceType;
}

export type RNLoggerOptions = {
  maxRequests?: number;
};

export interface ILog {
  _id: number;
  startTime: number;
  type: LogType;
}
