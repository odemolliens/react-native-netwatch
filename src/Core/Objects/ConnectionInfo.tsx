import { ILog, LogType } from '../../types';
import { NetInfoState } from '@react-native-community/netinfo/src/internal/types';

export class ConnectionInfo implements ILog {
  _id: number = -1;
  type: LogType = 'CONNECT';
  startTime: number = Date.now();
  connection: NetInfoState | null = null;

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ConnectionInfo;
