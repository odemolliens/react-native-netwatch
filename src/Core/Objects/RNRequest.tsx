import { ILog, LogType } from '../../types';
// @ts-ignore
// BlobFileReader is needed to read the content of the body response (it's type can be blob)
import BlobFileReader from 'react-native/Libraries/Blob/FileReader';

const fromEntries = (arr: any[]) =>
  arr.reduce((acc, [k, v]) => {
    acc[k] = v;
    return acc;
  }, {});

export const stringifyData = (data: any) => {
  try {
    if (data?._parts?.length) {
      return JSON.stringify(fromEntries(data?._parts), null, 2);
    }
    return JSON.stringify(JSON.parse(data), null, 2);
  } catch (e) {
    return `${data}`;
  }
};

export const getRequestBody = (dataSent: any): string => stringifyData(dataSent) || '';

export const getResponseBody = async (responseType: string, response?: any): Promise<string> => {
  if (!response) return '';
  const _responseBody = await (responseType !== 'blob' ? response : parseResponseBlob(response));
  return stringifyData(_responseBody || '');
};

export const parseResponseBlob = async (response: Blob) => {
  const blobReader = new BlobFileReader();
  blobReader.readAsText(response);

  return await new Promise<string>((resolve, reject) => {
    const handleError = () => reject(blobReader.error);

    blobReader.addEventListener('load', () => {
      resolve(blobReader.result);
    });
    blobReader.addEventListener('error', handleError);
    blobReader.addEventListener('abort', handleError);
  });
};

export class RNRequest implements ILog {
  _id: number = -1;
  type: LogType = 'RNR';
  startTime: number = Date.now();
  readyState: number = 0;
  url: string = '';
  shortUrl?: string = '';
  method: string = '';
  status: number = -1;
  endTime: number = 0;
  timeout?: number;
  dataSent: string = '';
  requestHeaders?: any;
  responseHeaders?: any;
  responseContentType?: string;
  responseSize?: number;
  responseType?: string;
  responseURL?: string;
  response: string = '';

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default RNRequest;
