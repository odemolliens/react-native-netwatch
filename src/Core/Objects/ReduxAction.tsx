import { ILog, LogType } from '../../types';

interface IAction {
  type: string;
  payload: any;
}

export class ReduxAction implements ILog {
  _id: number = -1;
  type: LogType = 'REDUX';
  startTime: number = Date.now();
  stringifiedAction: string = '';
  extra: string = '';
  action: IAction = { type: '__ERROR:UNDEFINED__', payload: '' };

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ReduxAction;
