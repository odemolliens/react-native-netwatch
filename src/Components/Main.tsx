import * as React from 'react';
import { useState, useEffect, useContext, useRef } from 'react';
import { View, StyleSheet, Alert, TouchableOpacity, FlatList, Share, Keyboard } from 'react-native';
import { Appbar, Searchbar } from 'react-native-paper';
import Item from './Item';
import FeatherIcon from 'react-native-vector-icons/Feather';
import { Settings } from './Settings';
import { ILog, SourceType, RequestMethod, EnumSourceType, EnumFilterType } from '../types';
import RNRequest from '../Core/Objects/RNRequest';
import ReduxAction from '../Core/Objects/ReduxAction';
import NRequest from '../Core/Objects/NRequest';
import { ThemeContext } from '../Theme';

export interface IProps {
  testId?: string;
  onPress: Function;
  onPressClose: (value: boolean) => void;
  onPressDetail: (value: boolean) => void;
  visible?: boolean;
  reduxActions: ReduxAction[];
  rnRequests: RNRequest[];
  nRequests: NRequest[];
  clearAll: Function;
  maxRequests?: number;
}

const _positionFlatList = 0;

export const Main = (props: IProps) => {
  const theme = useContext(ThemeContext);
  const refFlatList = useRef<FlatList<any>>(null);
  const [flatListLastOffset, setFlatListLastOffset] = useState<number>(0);
  const [requests, setRequests] = useState<Array<ILog>>([]);
  const [searchQuery, setSearchQuery] = useState<string>('');
  const [source, setSource] = useState<SourceType | EnumSourceType>(EnumSourceType.All);
  const [filter, setFilter] = useState<RequestMethod | EnumFilterType>(EnumFilterType.All);
  const [settingsVisible, setSettingsVisible] = useState(false);
  const [deleteVisible, setDeleteVisible] = useState<boolean>(false);
  const onChangeSearch = (query: string) => setSearchQuery(query);
  const hideDialogAndDelete = () => {
    clearList();
    if (deleteVisible) setDeleteVisible(false);
  };

  useEffect(() => {
    if (source === EnumSourceType.Redux) {
      if (filter !== EnumFilterType.All) setFilter(EnumFilterType.All);
      setRequests(props.reduxActions);
    } else if (source === EnumSourceType.ReactNativeRequest) {
      setRequests(props.rnRequests);
    } else if (source === EnumSourceType.Nativerequest) {
      setRequests(props.nRequests.sort(compare).reverse());
    } else {
      setRequests(mergeArrays(props.reduxActions, props.rnRequests, props.nRequests).sort(compare).reverse());
    }

    // TODO: Keep the same color after update for each items
    // TODO: We can add many items during one loop -> we must set the offset to handle that
    // TODO: What if we filtering -> scroll to 0 ?
    if (_positionFlatList > 55) {
      setFlatListLastOffset(_positionFlatList);
    }
  }, [props.rnRequests, props.reduxActions, props.nRequests, source, filter]);

  useEffect(() => {
    refFlatList?.current?.scrollToOffset({
      animated: false,
      offset: flatListLastOffset + 60 * 1,
    });
  }, [flatListLastOffset]);

  const mergeArrays = (...arrays: Array<ILog[]>) => {
    return [...arrays.flat()];
  };

  const compare = (a: ILog, b: ILog) => {
    const startTimeA = a.startTime;
    const startTimeB = b.startTime;

    let comparison = 0;
    if (startTimeA > startTimeB) {
      comparison = 1;
    } else if (startTimeA < startTimeB) {
      comparison = -1;
    }
    return comparison;
  };

  const filteredRequests = (): Array<any> =>
    requests.slice(0, props.maxRequests).filter(request => {
      if (filter === EnumFilterType.All) {
        if (request instanceof ReduxAction) {
          return JSON.stringify(request.action).toLowerCase().includes(searchQuery.toLowerCase());
        }
      }
      if (request instanceof RNRequest || request instanceof NRequest) {
        if (request.type === EnumSourceType.ReactNativeRequest) {
          if (filter === request.method || filter === EnumFilterType.All) {
            return (
              request.type === EnumSourceType.ReactNativeRequest &&
              (request.url?.toLowerCase().includes(searchQuery.toLowerCase()) ||
                request.status === parseInt(searchQuery, 10) ||
                request.method.toLowerCase() === searchQuery.toLowerCase())
            );
          }
        } else {
          if (filter === request.method || filter === EnumFilterType.All) {
            return (
              request.type === EnumSourceType.Nativerequest &&
              (request.url?.toLowerCase().includes(searchQuery.toLowerCase()) ||
                request.status === parseInt(searchQuery, 10) ||
                request.method.toLowerCase() === searchQuery.toLowerCase())
            );
          }
        }
      }
      return false;
    });

  const clearList = () => {
    props.clearAll();
    setRequests([]);
  };

  const formatSharedMessage = (array: Array<ILog>): string => {
    const _report = array.map(item => {
      return Object.values(item).flat().join(';');
    });
    return _report.join('\n');
  };

  const onShare = async (): Promise<void> => {
    try {
      await Share.share({
        message: formatSharedMessage(requests),
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  const onDelete = () => {
    Alert.alert('Warning', 'Do you really want clear the call list?', [
      {
        text: 'Cancel',
        onPress: () => setDeleteVisible(false),
        style: 'cancel',
      },
      { text: 'OK', onPress: hideDialogAndDelete },
    ]);
  };

  const _renderItems = (_props: { item: ReduxAction | RNRequest | NRequest; index: number }) => {
    return (
      <Item
        testID={`${_props.index}`}
        item={_props.item}
        onPress={() => {
          props.onPress(_props.item);
          props.onPressDetail(true);
        }}
        color={theme.gray800}
      />
    );
  };

  return (
    <>
      <Appbar.Header style={[styles.header, { backgroundColor: theme.gray900 }]}>
        <TouchableOpacity style={[styles.button, { borderLeftWidth: 0 }]}>
          <FeatherIcon
            name="x"
            color={theme.white}
            size={24}
            onPress={() => (settingsVisible ? setSettingsVisible(false) : props.onPressClose(false))}
          />
        </TouchableOpacity>
        <Appbar.Content color={theme.blue500} title="Netwatch" titleStyle={{ fontSize: 18 }} />
        <TouchableOpacity style={[styles.button, { borderLeftWidth: 0 }]}>
          <FeatherIcon name="download" color={theme.white} size={24} onPress={onShare} />
        </TouchableOpacity>
      </Appbar.Header>
      <View style={[styles.options, { backgroundColor: theme.gray800 }]}>
        <Searchbar
          onFocus={() => {
            setSettingsVisible(false);
          }}
          onChange={() => {
            setSettingsVisible(false);
          }}
          placeholderTextColor={theme.gray400}
          iconColor={theme.gray500}
          inputStyle={{ color: theme.gray400 }}
          style={[styles.searchBar, { backgroundColor: theme.gray800 }]}
          placeholder={'Search'}
          onChangeText={onChangeSearch}
          value={searchQuery}
        />

        <TouchableOpacity
          testID="showDetailsButton"
          style={[
            styles.button,
            {
              borderLeftColor: theme.gray900,
              backgroundColor: settingsVisible ? theme.gray900 : theme.gray800,
            },
          ]}
          onPress={() => {
            Keyboard.dismiss();
            setSettingsVisible(!settingsVisible);
          }}
        >
          <FeatherIcon
            name="filter"
            color={filter !== EnumFilterType.All || source !== EnumSourceType.All ? theme.blue500 : theme.gray300}
            size={24}
          />
        </TouchableOpacity>

        <TouchableOpacity
          testID="deleteListButton"
          style={[styles.button, { borderLeftColor: theme.gray900 }]}
          onPress={onDelete}
        >
          <FeatherIcon name="trash-2" color={theme.gray300} size={24} />
        </TouchableOpacity>
      </View>

      {settingsVisible ? (
        <Settings source={source} onSetSource={setSource} filter={filter} onSetFilter={setFilter} />
      ) : (
        <FlatList
          ref={refFlatList}
          maintainVisibleContentPosition={{
            autoscrollToTopThreshold: 10,
            minIndexForVisible: 1,
          }}
          style={{ backgroundColor: theme.gray800 }}
          renderItem={_renderItems}
          keyExtractor={item => `${item._id.toString()}${item.startTime}${item.type}`}
          data={filteredRequests()}
        />
      )}
    </>
  );
};

const styles = StyleSheet.create({
  header: {
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },
  searchBar: {
    flex: 1,
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },
  options: {
    height: 56,
    flexDirection: 'row',
    alignItems: 'center',
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
