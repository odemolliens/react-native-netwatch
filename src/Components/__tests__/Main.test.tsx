import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Main, IProps } from '../Main';
import ReduxAction from '../../Core/Objects/ReduxAction';
import { RNRequest } from '../../Core/Objects/RNRequest';
import { EnumFilterType, EnumSourceType } from '../../types';
import NRequest from '../../Core/Objects/NRequest';
import { Alert } from 'react-native';
import Share from 'react-native-share';

describe('Main test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  let useEffect;
  let mockUseEffect;
  const setSource = jest.fn();
  const setFilter = jest.fn();
  const setSearchQuery = jest.fn();
  const setRequests = jest.fn();
  const setSettingsVisible = jest.fn();
  const setDeleteVisible = jest.fn();
  const setLoadingXLSX = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect = () => {
      useEffect.mockImplementationOnce(f => f());
    };

    jest.mock('react', () => ({
      ...jest.requireActual('react'),
      useState: jest.fn(),
    }));
  });

  describe('Tests source', () => {
    it('should render properly', () => {
      component = shallow(<Main {...props} />);
      expect(component).toMatchSnapshot();
    });

    it('should run set source React native Request', () => {
      const useStateMock: any = (source: any) => [EnumSourceType.ReactNativeRequest, setSource];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setSource).toHaveBeenCalledTimes(1);
    });

    it('should run set source Redux', () => {
      const useStateMock: any = (source: any) => [EnumSourceType.Redux, setSource];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setSource).toHaveBeenCalledTimes(2);
    });

    it('should run set source Native Request', () => {
      const useStateMock: any = (source: any) => [EnumSourceType.Nativerequest, setSource];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setSource).toHaveBeenCalledTimes(1);
    });

    it('should run set source All', () => {
      const useStateMock: any = (source: any) => [EnumSourceType.All, setSource];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setSource).toHaveBeenCalledTimes(1);
    });
  });

  describe('Tests filter', () => {
    it('should run set filter All', () => {
      const useStateMock: any = (source: any) => [EnumFilterType.All, setFilter];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setFilter).toHaveBeenCalledTimes(1);
    });

    it('should run set filter Get', () => {
      const useStateMock: any = (source: any) => [EnumFilterType.Get, setFilter];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setFilter).toHaveBeenCalledTimes(1);
    });
  });

  describe('Tests search', () => {
    it('should run set search', () => {
      const useStateMock: any = (source: any) => ['200', setSearchQuery];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setSearchQuery).toHaveBeenCalledTimes(1);
    });
  });

  describe('Tests requests', () => {
    it('should run set requests', () => {
      const useStateMock: any = (source: any) => [[], setRequests];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setRequests).toHaveBeenCalledTimes(1);
    });
  });

  describe('Tests share XLSX', () => {
    it('should run set loadingXLSX', () => {
      jest.spyOn(Share, 'open');
      const useStateMock: any = (loadingXLSX: any) => [true, setLoadingXLSX];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      mockUseEffect();
      givenProps();
      givenComponent();
      expect(setLoadingXLSX).toHaveBeenCalledTimes(1);
      // expect(Share.open).toHaveBeenCalledTimes(1);
    });
  });

  it('should render properly and showSettings', () => {
    const useStateMock: any = (settingsVisible: any) => [settingsVisible, setSettingsVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps();
    givenComponent();
    expect(component).toMatchSnapshot();
    whenPressingButton('showSettingsButton');
    expect(setSettingsVisible).toHaveBeenCalledTimes(1);
    expect(setSettingsVisible).toHaveBeenCalledWith(true);
  });

  it('should render properly and delete list', () => {
    const useStateMock: any = (deleteVisible: any) => [deleteVisible, setDeleteVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const spyAlert = jest.spyOn(Alert, 'alert');
    givenProps();
    givenComponent();
    // component = shallow(
    //   <Main reduxActions={mockActions} rnRequests={mockRNRequests} nRequests={mockNRequests} {...props} />,
    // );
    component.find(`[testID="itemsList"]`).props().keyExtractor({ _id: 73 });
    component.find(`[testID="itemsList"]`).props().getItemLayout({ _data: null, index: 73 });
    component.find(`[testID="itemsList"]`).props().renderItem({ item: mockActions[0], index: 73 });
    // expect(component.find('FlatList').prop('data')).not.toStrictEqual([])

    component.find('[testID="deleteListButton"]').simulate('press');
    expect(Alert.alert).toHaveBeenCalledTimes(1);
    // Press cancel button
    // @ts-ignore
    spyAlert.mock.calls[0][2][0].onPress();

    component.find('[testID="deleteListButton"]').simulate('press');
    expect(Alert.alert).toHaveBeenCalledTimes(2);
    // Press ok button
    // @ts-ignore
    spyAlert.mock.calls[0][2][1].onPress();

    expect(props.clearAll).toHaveBeenCalledTimes(1);
  });

  it('should render properly and close', () => {
    // To close Netwatch, settingsVisible must be false, if not, press x will close settings screen and not Netwatch
    const useStateMock: any = (settingsVisible: any) => [false, setSettingsVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps();
    givenComponent();
    component.find('[name="x"]').simulate('press');
    expect(props.onPressClose).toHaveBeenCalledTimes(1);
    expect(props.onPressClose).toHaveBeenCalledWith(false);
  });

  // UTILITIES
  function whenPressingButton(testId: string) {
    component.find(`[testID="${testId}"]`).simulate('press');
  }

  // GIVEN
  function givenComponent() {
    component = shallow(<Main {...props} />);
  }

  function givenProps(visible: boolean = true) {
    props = {
      testId: 'mainScreen',
      onPress: jest.fn(),
      onPressClose: jest.fn(),
      onPressDetail: jest.fn(),
      reduxActions: mockActions,
      rnRequests: mockRNRequests,
      nRequests: mockNRequests,
      clearAll: jest.fn(),
      maxRequests: 50,
    };
  }
});

const mockActions: ReduxAction[] = [
  new ReduxAction({
    _id: 73,
    startTime: 100,
    type: 'REDUX',
    stringifiedAction: "{ type: '__ERROR:UNDEFINED__', payload: '' }",
    action: { type: '__ERROR:UNDEFINED__', payload: '' },
  }),
];

const mockRNRequests: RNRequest[] = [
  new RNRequest({
    _id: 74,
    dataSent: 'dataSent',
    endTime: 1613477575757,
    method: 'GET',
    readyState: 4,
    response: 'response',
    responseContentType: 'application/json',
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 200,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  }),
];

const mockNRequests: NRequest[] = [
  new NRequest({
    _id: 75,
    dataSent: 'dataSent',
    endTime: 1613477575757,
    method: 'GET',
    readyState: 4,
    response: 'response',
    responseContentType: 'application/json',
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 200,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  }),
];
