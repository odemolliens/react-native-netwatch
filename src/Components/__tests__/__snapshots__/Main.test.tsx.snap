// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main test suite Tests source should render properly 1`] = `
<Fragment>
  <Header
    style={
      Array [
        Object {
          "elevation": 0,
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
        },
        Object {
          "backgroundColor": "#111827",
        },
      ]
    }
  >
    <ForwardRef
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
    >
      <Icon
        color="#F9FAFB"
        name="x"
        onPress={[Function]}
        size={24}
      />
    </ForwardRef>
    <Content
      color="#06B6D4"
      title="Netwatch"
      titleStyle={
        Object {
          "fontSize": 18,
        }
      }
    />
    <ForwardRef
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
    >
      <Icon
        color="#F9FAFB"
        name="download"
        onPress={[Function]}
        size={24}
      />
    </ForwardRef>
  </Header>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 56,
        },
        Object {
          "backgroundColor": "#1F2937",
        },
      ]
    }
  >
    <Searchbar
      iconColor="#6B7280"
      inputStyle={
        Object {
          "color": "#9CA3AF",
        }
      }
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      placeholderTextColor="#9CA3AF"
      style={
        Array [
          Object {
            "elevation": 0,
            "flex": 1,
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
          },
          Object {
            "backgroundColor": "#1F2937",
          },
        ]
      }
      value=""
    />
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "backgroundColor": "#1F2937",
            "borderLeftColor": "#111827",
          },
        ]
      }
      testID="showSettingsButton"
    >
      <Icon
        color="#D1D5DB"
        name="filter"
        size={24}
      />
    </ForwardRef>
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftColor": "#111827",
          },
        ]
      }
      testID="deleteListButton"
    >
      <Icon
        color="#D1D5DB"
        name="trash-2"
        size={24}
      />
    </ForwardRef>
  </Component>
  <FlatList
    data={Array []}
    disableVirtualization={false}
    getItemLayout={[Function]}
    horizontal={false}
    initialNumToRender={20}
    keyExtractor={[Function]}
    maintainVisibleContentPosition={
      Object {
        "autoscrollToTopThreshold": 10,
        "minIndexForVisible": 1,
      }
    }
    maxToRenderPerBatch={30}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
    testID="itemsList"
    updateCellsBatchingPeriod={70}
    windowSize={21}
  />
</Fragment>
`;

exports[`Main test suite should render properly and showSettings 1`] = `
<Fragment>
  <Header
    style={
      Array [
        Object {
          "elevation": 0,
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
        },
        Object {
          "backgroundColor": "#111827",
        },
      ]
    }
  >
    <ForwardRef
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
    >
      <Icon
        color="#F9FAFB"
        name="x"
        onPress={[Function]}
        size={24}
      />
    </ForwardRef>
    <Content
      color="#06B6D4"
      title="Netwatch"
      titleStyle={
        Object {
          "fontSize": 18,
        }
      }
    />
    <ForwardRef
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
    >
      <Icon
        color="#F9FAFB"
        name="download"
        onPress={[Function]}
        size={24}
      />
    </ForwardRef>
  </Header>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 56,
        },
        Object {
          "backgroundColor": "#1F2937",
        },
      ]
    }
  >
    <Searchbar
      iconColor="#6B7280"
      inputStyle={
        Object {
          "color": "#9CA3AF",
        }
      }
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      placeholderTextColor="#9CA3AF"
      style={
        Array [
          Object {
            "elevation": 0,
            "flex": 1,
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
          },
          Object {
            "backgroundColor": "#1F2937",
          },
        ]
      }
      value=""
    />
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "backgroundColor": "#1F2937",
            "borderLeftColor": "#111827",
          },
        ]
      }
      testID="showSettingsButton"
    >
      <Icon
        color="#D1D5DB"
        name="filter"
        size={24}
      />
    </ForwardRef>
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftColor": "#111827",
          },
        ]
      }
      testID="deleteListButton"
    >
      <Icon
        color="#D1D5DB"
        name="trash-2"
        size={24}
      />
    </ForwardRef>
  </Component>
  <FlatList
    data={Array []}
    disableVirtualization={false}
    getItemLayout={[Function]}
    horizontal={false}
    initialNumToRender={20}
    keyExtractor={[Function]}
    maintainVisibleContentPosition={
      Object {
        "autoscrollToTopThreshold": 10,
        "minIndexForVisible": 1,
      }
    }
    maxToRenderPerBatch={30}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
    testID="itemsList"
    updateCellsBatchingPeriod={70}
    windowSize={21}
  />
</Fragment>
`;
